examples/gdp/jobshop.py

```diff
@@ -29,51 +29,64 @@
-def _L_filter(model, I, K, J):
-    return I < K and model.tau[I,J] and model.tau[K,J]
-model.L = Set( initialize=model.JOBS * model.JOBS * model.STAGES,
-               dimen=3, filter=_L_filter)
+    # Auto-generate the L set (potential collisions between 2 jobs at any stage.
+    def _L_filter(model, I, K, J):
+        return I < K and model.tau[I,J] and model.tau[K,J]
+    model.L = Set( initialize=model.JOBS * model.JOBS * model.STAGES,
+                   dimen=3, filter=_L_filter)
 
-# Makespan is greater than the start time of every job + that job's
-# total duration
-def _feas(model, I):
-    return model.ms >= model.t[I] + sum(model.tau[I,M] for M in model.STAGES)
-model.Feas = Constraint(model.JOBS, rule=_feas)
+    # Makespan is greater than the start time of every job + that job's
+    # total duration
+    def _feas(model, I):
+        return model.ms >= model.t[I] + sum(model.tau[I,M] for M in model.STAGES)
+    model.Feas = Constraint(model.JOBS, rule=_feas)
 
-# Disjunctions to prevent clashes at a stage: This creates a set of
-# disjunct pairs: one if job I occurs before job K and the other if job
-# K occurs before job I.
-def _NoClash(disjunct, I, K, J, IthenK):
-    model = disjunct.model()
-    lhs = model.t[I] + sum([M<J and model.tau[I,M] or 0 for M in model.STAGES])
-    rhs = model.t[K] + sum([M<J and model.tau[K,M] or 0 for M in model.STAGES])
-    if IthenK:
-        disjunct.c = Constraint(expr=lhs+model.tau[I,J]<=rhs)
-    else:
-        disjunct.c = Constraint(expr=rhs+model.tau[K,J]<=lhs)
-model.NoClash = Disjunct(model.L, model.I_BEFORE_K, rule=_NoClash)
+    # Disjunctions to prevent clashes at a stage: This creates a set of
+    # disjunct pairs: one if job I occurs before job K and the other if job
+    # K occurs before job I.
+    def _NoClash(disjunct, I, K, J, IthenK):
+        model = disjunct.model()
+        lhs = model.t[I] + sum([M<J and model.tau[I,M] or 0 for M in model.STAGES])
+        rhs = model.t[K] + sum([M<J and model.tau[K,M] or 0 for M in model.STAGES])
+        if IthenK:
+            disjunct.c = Constraint(expr=lhs+model.tau[I,J]<=rhs)
+        else:
+            disjunct.c = Constraint(expr=rhs+model.tau[K,J]<=lhs)
+    model.NoClash = Disjunct(model.L, model.I_BEFORE_K, rule=_NoClash)
 
-# Define the disjunctions: either job I occurs before K or K before I
-def _disj(model, I, K, J):
-    return [model.NoClash[I,K,J,IthenK] for IthenK in model.I_BEFORE_K]
-model.disj = Disjunction(model.L, rule=_disj)
+    # Define the disjunctions: either job I occurs before K or K before I
+    def _disj(model, I, K, J):
+        return [model.NoClash[I,K,J,IthenK] for IthenK in model.I_BEFORE_K]
+    model.disj = Disjunction(model.L, rule=_disj)
 
-# minimize makespan
-model.makespan = Objective(expr=model.ms)
+    # minimize makespan
+    model.makespan = Objective(expr=model.ms)
+    return model
+
+
+def build_small_concrete():
+    return build_model().create_instance('jobshop-small.dat')
```

Comment:
Why make this a hard-coded function?  It seems to only be called in the `if __name__ == '__main__'` below.  I think it can be removed.

Reply:
I added these in here to associate the abstract model files with the `.dat` files that accompany them. Otherwise you're left with an abstract model and you'd be hard-coding the `dat` association somewhere else.